### PR TITLE
[refactor] Store and restore state along with checkpoints

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

This PR adds a mechanism to store data that is then written out as a JSON file (`training_status.json`) that can be loaded on resume. This is done through a new class (`GlobalTrainingStatus`) that keeps both metadata bout the JSON file and the key/values that need to be written, organized by behavior name. Currently, it stores just the lesson number during curriculum. This value is loaded when `--resume` is specified. 

`training_status.json` is versioned in a very similar way to the `timers.json` file, and versions are checked on resume. Warnings are thrown if the version doesn't match. 

We also no longer need the `--lesson` CLI option as that was used to reset a lesson on resume. 

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [x] Code refactor
- [ ] Breaking change
- [x] Documentation update
- [ ] Other (please describe)

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: com.unity.ml-agents/CHANGELOG.md [hunk 1]
```diff
@@ -17,18 +17,20 @@ and this project adheres to
...
 - The format for trainer configuration has changed, and the "default" behavior has been deprecated.
   See the [Migration Guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) for more details. (#3936)
 - Training artifacts (trained models, summaries) are now found in the `results/`
   directory. (#3829)
+- When using Curriculum, the current lesson will resume if training is quit and resumed. As such,
+  the `--lesson` CLI option has been removed. (#4025)
+### Minor Changes
+#### com.unity.ml-agents (C#)
+- `ObservableAttribute` was added. Adding the attribute to fields or properties on an Agent will allow it to generate
+  observations via reflection.
```
Comment (chriselion) [inline com.unity.ml-agents/CHANGELOG.md:32]:
```suggestion
  observations via reflection. (#3925, #4006)
```
(bad merge conflict resolution?)

File: ml-agents/mlagents/trainers/stats.py [hunk 2]
```diff
@@ -304,6 +343,39 @@ def __init__(self, category: str):
     def add_writer(writer: StatsWriter) -> None:
         StatsReporter.writers.append(writer)
 
+    @staticmethod
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/stats.py:346]: StatsReporter doesn't seem like it's the right place to be saving and loading global state. Can you make a new class for this? TrainingGlobalState?

File: ml-agents/mlagents/trainers/learn.py [hunk 3]
```diff
@@ -82,6 +82,9 @@ def run_training(run_seed: int, options: RunOptions) -> None:
         )
         # Make run logs directory
         os.makedirs(run_logs_dir, exist_ok=True)
+        # Load any needed states
+        if checkpoint_settings.resume:
+            StatsReporter.load_state(os.path.join(run_logs_dir, "training_status.json"))
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/learn.py:87]: Move `"training_status.json"` to a constant?

File: ml-agents/mlagents/trainers/meta_curriculum.py [hunk 4]
```diff
@@ -115,16 +116,24 @@ def increment_lessons(self, measure_vals, reward_buff_sizes=None):
...
+    def try_restore_all_curriculum(self):
         """
-        for _, curriculum in self.brains_to_curricula.items():
-            curriculum.lesson_num = lesson_num
+        Tries to restore all the curriculums to what is saved in training_status.json
+        """
+
+        for brain_name, curriculum in self.brains_to_curricula.items():
+            # Create a temporary StatsReporter with the right brain name
+            _statsreporter = StatsReporter(brain_name)
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/meta_curriculum.py:126]: It's unclear what's going on here; it feels really hacky (and probably brittle).

File: ml-agents/mlagents/trainers/stats.py [hunk 5]
```diff
@@ -304,6 +343,39 @@ def __init__(self, category: str):
...
+        try:
+            with open(path, "r") as f:
+                loaded_dict = json.load(f)
+            # Compare the metadata
+            _metadata = loaded_dict[StatsPropertyType.STATS_METADATA.value]
+            StatsMetaData.from_dict(_metadata).check_compatibility(StatsMetaData())
+            # Update saved state.
+            StatsReporter.saved_state.update(loaded_dict)
+        except FileNotFoundError:
+            pass
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/stats.py:361]: Should there be a warning here, or is it expected that this won't be there most of the time?

File: ml-agents/mlagents/trainers/learn.py [hunk 6]
```diff
@@ -159,6 +162,7 @@ def run_training(run_seed: int, options: RunOptions) -> None:
         env_manager.close()
         write_run_options(write_path, options)
         write_timing_tree(run_logs_dir)
+        StatsReporter.save_state(os.path.join(run_logs_dir, "training_status.json"))
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/learn.py:165]: I don't think this is the best place to store this global state. I would put it within the save model calls.

File: ml-agents/mlagents/trainers/stats.py (hunk 2 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/stats.py:346]: The main reason I put it in StatsReporter was that the trainers already have one initialized with the right key (the brain_name). No problem, it can be moved

File: ml-agents/mlagents/trainers/learn.py (hunk 6 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/learn.py]: Hmm, that makes sense since we'd want to resume on the event of a crash as well. 

However the save model calls will be moved to within the trainers, since it seems logical that each trainer drives its own checkpointing (at possibly different frequencies). Wouldn't want the global save to be there as well. We could have each trainer manage its own save state, or have every trainer just trigger the save_state function.

An alternative would be to just write to the JSON every time a new state is written. For stuff like the lesson number (which is very infrequently written) this is OK.

File: ml-agents/mlagents/trainers/stats.py (hunk 5 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/stats.py]: Added a warning. This should only happen if the user is loading from an older version of ML-Agents that did not save out the `training_status.json` file, or (as Vince mentioned below) the file was not saved out due to a crash.

File: ml-agents/mlagents/trainers/learn.py (hunk 3 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/learn.py]: I've moved it to a constant and the method to a helper method similar to the timing tree and configuration file writes.

File: ml-agents/mlagents/trainers/meta_curriculum.py (hunk 4 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/meta_curriculum.py]: It's less of a concern now since we're no longer using StatsReporters, but we're still using the `brain_name` to refer to meta curriculums.

File: ml-agents/mlagents/trainers/training_status.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    saved_state: Dict[str, Dict[str, Any]] = defaultdict(lambda: {})
+
+    def __init__(self, category: str):
+        """
+        Generic GlobalTrainingStatus writer. A category is the broadest type of storage (would
+        correspond the run name and trainer name, e.g. 3DBalltest_3DBall. A key is the
+        type of status needed to be saved (e.g. Lesson Number). Finally the Value is the float value
+        attached to this stat.
+        """
+        self.category: str = category
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/training_status.py:62]: This still feels awkward having a mix of instance and static data. Why not drop `self.category` and add an extra argument to `restore_parameter_state` and `store_parameter_state`?

File: ml-agents/mlagents/trainers/training_status.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        Save a JSON file that contains saved state.
+        :param path: Path to the JSON file containing the state.
+        """
+        GlobalTrainingStatus.saved_state[
+            StatusType.STATS_METADATA.value
+        ] = StatusMetaData().to_dict()
+        with open(path, "w") as f:
+            json.dump(GlobalTrainingStatus.saved_state, f, indent=4)
+
+    def store_parameter_state(self, key: StatusType, value: Any) -> None:
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/training_status.py:99]: nit: `set_parameter_state()` and `get_paramater_state()`? I think `restore_parameter_state` is a bad name since "restore" makes it sound like it's doing loading.

File: ml-agents/mlagents/trainers/training_status.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        """
+        Stores an arbitrary-named parameter in the global saved state.
+        :param key: The parameter, e.g. lesson number.
+        :param value: The value.
+        """
+        GlobalTrainingStatus.saved_state[self.category][key.value] = value
+
+    def restore_parameter_state(self, key: StatusType) -> Any:
+        """
+        Stores an arbitrary-named parameter in training_status.json.
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/training_status.py:109]: copy-pasted docstring.

File: ml-agents/mlagents/trainers/tests/test_training_status.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        test_json = json.load(fp)
+
+    assert "Category1" in test_json
+    assert StatusType.LESSON_NUM.value in test_json["Category1"]
+    assert test_json["Category1"][StatusType.LESSON_NUM.value] == 3
+    assert "metadata" in test_json
+
+    statsreporter_new = GlobalTrainingStatus("Category1")
+    GlobalTrainingStatus.load_state(path_dir)
+    restored_val = statsreporter_new.restore_parameter_state(StatusType.LESSON_NUM)
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/tests/test_training_status.py:29]: Also test that restore_parameter_state() on an unknown category or StatusType returns None instead of raising an exception.

File: ml-agents/mlagents/trainers/training_status.py (hunk 7 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/training_status.py:62]: I went this way to keep it inline with the StatsReporter, and so that if a trainer uses it multiple times it doesn't need to continue to pass in the category. 

Don't have a strong preference since it's used much less frequently than StatsReporter; I'm OK with changing it to a parameter to the class methods.

File: ml-agents/mlagents/trainers/training_status.py (hunk 9 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/training_status.py:109]: 👍 Fixed

File: ml-agents/mlagents/trainers/training_status.py (hunk 8 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/training_status.py]: Changed.

File: ml-agents/mlagents/trainers/training_status.py (hunk 7 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/training_status.py:62]: Changed it to two args for `get_parameter_state` and `set_parameter_state`.

File: ml-agents/mlagents/trainers/tests/test_training_status.py (hunk 10 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/tests/test_training_status.py]: Added test for these.

File: ml-agents/mlagents/trainers/training_status.py (hunk 7 above)
Comment (chriselion) [reply ml-agents/mlagents/trainers/training_status.py:62]: Can you remove self.category (and the initializer) now?

Comment (ervteng) [reply ml-agents/mlagents/trainers/training_status.py:62]: Yes - oversight on my part. Removed both.
